pkg/monitoring/mimir/service.go

```diff
@@ -0,0 +1,124 @@
+package mimir
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/observability-operator/pkg/monitoring"
+	"github.com/giantswarm/observability-operator/pkg/monitoring/mimir/ingress"
+)
+
+type MimirService struct {
+	client.Client
+}
+
+const (
+	ingressSecretName      = "mimir-gateway-ingress"
+	ingressSecretNamespace = "mimir"
+)
+
+func (ms *MimirService) ReconcileMimirConfig(ctx context.Context, mc string) error {
+	logger := log.FromContext(ctx).WithValues("cluster", mc)
+	logger.Info("ensuring mimir config")
+
+	err := ms.CreateOrUpdateIngressSecret(ctx, mc, logger)
+	if err != nil {
+		logger.Error(err, "failed to create or update mimir config")
+		return errors.WithStack(err)
+	}
+
+	logger.Info("ensured mimir config")
+
+	return nil
+}
+
+func (ms *MimirService) CreateOrUpdateIngressSecret(ctx context.Context, mc string, logger logr.Logger) error {
+	objectKey := client.ObjectKey{
+		Name:      ingressSecretName,
+		Namespace: ingressSecretNamespace,
+	}
+
+	current := &corev1.Secret{}
+	err := ms.Client.Get(ctx, objectKey, current)
+	if apierrors.IsNotFound(err) {
+		// CREATE SECRET
+		logger.Info("building ingress secret")
+
+		password, err := GetMimirIngressPassword(ctx, mc)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		secret, err := ingress.BuildIngressSecret(mc, password)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		err = ms.Client.Create(ctx, secret)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		logger.Info("ingress secret successfully created")
+
+		return nil
+	} else if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// UPDATE SECRET
```

Comment:
Do we need the upgrade part?

Reply:
I guess we don't since in any case with the usage of `htpasswd` command we can't update the secret's value as a new hash will be generated each time.